source/server/admin/clusters_chunk_processor.h

```diff
@@ -0,0 +1,131 @@
+  virtual ~ClustersChunkProcessor() = default;
+};
+
+/**
+ * TextClustersChunkProcessor streams the response using a newline delimited text format.
+ */
+class TextClustersChunkProcessor : public ClustersChunkProcessor {
+public:
+  TextClustersChunkProcessor(uint64_t chunk_limit, Http::ResponseHeaderMap& response_headers,
+                             const Upstream::ClusterManager::ClusterInfoMap& cluster_info_map);
+  bool nextChunk(Buffer::Instance& response) override;
+
+private:
+  void render(std::reference_wrapper<const Upstream::Cluster> cluster, Buffer::Instance& response);
+
+  static void addOutlierInfo(const std::string& cluster_name,
+                             const Upstream::Outlier::Detector* outlier_detector,
+                             Buffer::Instance& response);
+  static void addCircuitBreakerSettings(const std::string& cluster_name,
+                                        const std::string& priority_str,
+                                        Upstream::ResourceManager& resource_manager,
+                                        Buffer::Instance& response);
+  const uint64_t chunk_limit_;
+  Http::ResponseHeaderMap& response_headers_;
+  std::vector<std::reference_wrapper<const Upstream::Cluster>> clusters_;
+  uint64_t idx_;
+};
+
+/**
+ * ClustersJsonContext holds an Envoy::Json::Streamer and the top-level JSON objects throughout the
+ * duration of a request. When it is destructed, the buffer will terminate the array and object
+ * herein. See the Envoy::Json::Streamer implementation for details.
+ */
+struct ClustersJsonContext {
+  ClustersJsonContext(std::unique_ptr<Json::Streamer> streamer,
+                      std::reference_wrapper<Buffer::Instance> buffer,
+                      Json::Streamer::MapPtr root_map, Json::Streamer::ArrayPtr clusters)
+      : streamer_(std::move(streamer)), buffer_(buffer.get()), root_map_(std::move(root_map)),
+        clusters_(std::move(clusters)) {}
+  std::unique_ptr<Json::Streamer> streamer_;
+  Buffer::Instance& buffer_;
+  Json::Streamer::MapPtr root_map_;
+  Json::Streamer::ArrayPtr clusters_;
+};
+
+/**
+ * JsonClustersChunkProcessor streams the response and emulates the JSON encoding behavior of
+ * the gRPC clients; specifically, "zero-value" values such as false, 0, empty string, and
+ * empty objects are omitted from the output.
+ */
+class JsonClustersChunkProcessor : public ClustersChunkProcessor {
+public:
+  JsonClustersChunkProcessor(uint64_t chunk_limit, Http::ResponseHeaderMap& response_headers,
+                             const Upstream::ClusterManager::ClusterInfoMap& cluster_info_map);
+  bool nextChunk(Buffer::Instance& response) override;
+
+private:
+  void render(std::reference_wrapper<const Upstream::Cluster> cluster, Buffer::Instance& response);
+  void drainBufferIntoResponse(Buffer::Instance& response);
+  void finalize(Buffer::Instance& response);
+  void addAddress(Json::Streamer::Map* raw_host_ptr, const Upstream::HostSharedPtr& host,
```

Comment:
So, the constructors for json maps and arrays in `Envoy::Json::Streamer` create unique pointers by the name `Envoy::Json::Streamer::MapPtr` and `Envoy::Json::Streamer::ArrayPtr`. Within the code, sometimes I need to pass the map or array to a private helper function to do extra work without terminating the array or map (termination happens during deconstruction). By passing the raw pointer, I can loan the pointer to the map or array to the helper function to do work without it owning it.

I am open to recommendations; this just seemed like the most straight forward way to achieve this. If you think this is sensible I am happy to add comments to each of these private member functions that take the raw pointer.

Reply:
totally reasonable, but just for consistency (and some evidence that the caller thinks it should be non-null), can you pass by ref, and at call-site if you have a MapPtr you can just pass `*map`.

TBH what you have is what I would more naturally write, but Envoy has been using refs for this use-case since I started working on it.

Reply:
Ah, yeah. That actually makes sense. We do expect a value so having the helper member functions take a ref is a good approach. I'll try that.